### PR TITLE
resolve sanity check last id file acces problems

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #5156

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

- set inteval to 10 mins
- let it run for several days.
- reset last check id to 1
- run cli sanity check
- rerun cli sanity check
- let it run for several more days
-

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 1]
```diff
@@ -69,48 +69,51 @@ protected boolean checkItemCountByPstmt(CheckCase checkCase) throws SQLException
...
-            if(checkCase.checkId) {
-                pstmt.setInt(1, lastId);
-                pstmt.setInt(2, maxId);
-            }
-            try(ResultSet rs = pstmt.executeQuery();) {
-                if (rs.next() && (rs.getInt(1) > 0)) {
-                    errors.append(String.format("Error: Found %s %s\n", rs.getInt(1), checkCase.itemName));
-                    checkOk = false;
+        try (PreparedStatement pstmt = conn.prepareStatement(checkCase.getSqlTemplate())) {
+            if(checkCase.isCheckId()) {
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:73]:
```suggestion
            if (checkCase.isCheckId()) {
```

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 2]
```diff
@@ -69,48 +69,51 @@ protected boolean checkItemCountByPstmt(CheckCase checkCase) throws SQLException
...
         {
-            s_logger.error("checkItemCountByPstmt:Exception:"+e.getMessage());
-            throw new CloudRuntimeException("checkItemCountByPstmt:Exception:"+e.getMessage(),e);
+            throwPreparedStatementExcecutionException("preparing statement", checkCase.getSqlTemplate(), e);
+        }
+        return checkOk;
+    }
+
+    private boolean isCheckOkForPstmt(CheckCase checkCase, boolean checkOk, PreparedStatement pstmt) {
+        try(ResultSet rs = pstmt.executeQuery();) {
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:91]:
```suggestion
        try (ResultSet rs = pstmt.executeQuery();) {
```

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 3]
```diff
@@ -69,48 +69,51 @@ protected boolean checkItemCountByPstmt(CheckCase checkCase) throws SQLException
...
+            if (rs.next() && (rs.getInt(1) > 0)) {
+                errors.append(String.format("Error: Found %s %s%n", rs.getInt(1), checkCase.getItemName()));
+                checkOk = false;
+            }
+        }catch (Exception e)
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:96]:
```suggestion
        } catch (Exception e)
```

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 4]
```diff
@@ -120,6 +123,21 @@ protected void checkMaxUsage() throws SQLException {
...
+        try(ResultSet rs = pstmt.executeQuery();) {
+           if (rs.next()) {
+                aggregationRange = rs.getInt(1);
+            } else {
+               if (s_logger.isDebugEnabled()) {
+                   s_logger.debug("Failed to retrieve aggregation range. Using default : " + aggregationRange);
+               }
+            }
+        }catch (SQLException e) {
+            throwPreparedStatementExcecutionException("retrieval aggregate value is failing", pstmt.toString(), e);
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:136]:
```suggestion
            throwPreparedStatementExcecutionException("Retrieval aggregate value is failing", pstmt.toString(), e);
```

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 5]
```diff
@@ -170,14 +188,21 @@ protected void checkSnapshotUsage() {
...
         try(BufferedReader reader = new BufferedReader(new FileReader(lastCheckFile));) {
             String lastIdText = null;
             lastId = -1;
-            if ((reader != null) && (lastIdText = reader.readLine()) != null) {
+            if ((lastIdText = reader.readLine()) != null) {
                 lastId = Integer.parseInt(lastIdText);
             }
         } catch (Exception e) {
-            s_logger.error("readLastCheckId:Exception:"+e.getMessage(),e);
+            String msg = String.format("error reading the LastCheckId reason: %s", e.getMessage());
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:201]:
```suggestion
            String msg = String.format("Error reading the LastCheckId reason: %s", e.getMessage());
```

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 6]
```diff
@@ -170,14 +188,21 @@ protected void checkSnapshotUsage() {
...
+            s_logger.error(msg);
+            s_logger.debug(msg, e);
+        } finally {
+            s_logger.info(String.format("using %d as last checked id to start from in sanity check", lastId));
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:205]:
```suggestion
            s_logger.info(String.format("Using %d as last checked id to start from in sanity check", lastId));
```

other logs messages if applicable

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 7]
```diff
@@ -69,48 +69,51 @@ protected boolean checkItemCountByPstmt(CheckCase checkCase) throws SQLException
...
+        String msg = String.format("%s for prepared statement \"%s\" reason: %s", msgPrefix, stmt, e.getMessage());
+        s_logger.error(msg);
+        throw new CloudRuntimeException(msg, e);
+    }
+
     protected void checkMaxUsage() throws SQLException {
         int aggregationRange = DEFAULT_AGGREGATION_RANGE;
-        try (PreparedStatement pstmt = conn.prepareStatement(
-                "SELECT value FROM `cloud`.`configuration` where name = 'usage.stats.job.aggregation.range'");)
+        String sql = "SELECT value FROM `cloud`.`configuration` where name = 'usage.stats.job.aggregation.range'";
```
Comment (sureshanaparti) [inline usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:111]:
```suggestion
        String sql = "SELECT value FROM `cloud`.`configuration` WHERE name = 'usage.stats.job.aggregation.range'";
```

**Review by sureshanaparti — DISMISSED** [review]

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java [hunk 8]
```diff
@@ -170,14 +188,21 @@ protected void checkSnapshotUsage() {
     }
 
     protected void readLastCheckId(){
+        if (s_logger.isDebugEnabled()) {
+            s_logger.debug("reading last checked id for sanity check");
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:192]: This is excessive capitalisation to me, I won´t apply this.

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java (hunk 5 above)
Comment (DaanHoogland) [reply usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:201]: This is excessive capitalisation to me, I won´t apply.

File: usage/src/main/java/com/cloud/usage/UsageSanityChecker.java (hunk 6 above)
Comment (DaanHoogland) [reply usage/src/main/java/com/cloud/usage/UsageSanityChecker.java:205]: This is excessive capitalisation to me, I won´t apply.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@sureshanaparti are you fine with the changes?

**Review by nvazquez — APPROVED** [review]

**Review by rohityadavcloud — APPROVED** [review]

**Comment by vladimirpetrov** [comment]
The file **sanity-check-last-id** is still not updated when the usage jobs completes. It updated only when the cli sanity check is executed.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
> The file **sanity-check-last-id** is still not updated when the usage jobs completes. It updated only when the cli sanity check is executed.

can you have another look @vladimirpetrov I think the problem is the wait time, It will only start executing the sanity check after a day.

**Pull request merged**